Line Finding and Parameter Estimation (#362)
astropy/specutils: docs/fitting.rst

yielding a single composite model result (which can be split back into its
components if desired).

Line Finding
------------

There are two methods implemented in order to find emission and/or absorption
lines in a `~specutils.Spectrum1D` spectrum.

> The first method is `~specutils.fitting.find_lines_threshold` that will

Author: changed all method to technique

      ----------------- --------- -----------------
      4.572864321608041  emission                91
      4.824120603015076  emission                96
      5.477386934673367  emission               109
       8.99497487437186  emission               179

   >>> absorption_lines = lines[lines['line_type'] == 'absorption'] #doctest:+SKIP
         line_center    line_type  line_center_index
              um
      ----------------- ---------- -----------------
      8.190954773869347 absorption               163

An example using the `~specutils.fitting.find_lines_derivative`:

.. code-block:: python

   >>> import numpy as np
   >>> from astropy.modeling import models
   >>> import astropy.units as u
   >>> from specutils import Spectrum1D, SpectralRegion
   >>> from specutils.manipulation import noise_region_uncertainty
   >>> from specutils.fitting import find_lines_derivative

   >>> # Define the spectrum
   >>> np.random.seed(42)
   >>> g1 = models.Gaussian1D(1, 4.6, 0.2)
   >>> g2 = models.Gaussian1D(2.5, 5.5, 0.1)
   >>> g3 = models.Gaussian1D(-1.7, 8.2, 0.1)
   >>> x = np.linspace(0, 10, 200)
   >>> y = g1(x) + g2(x) + g3(x) + np.random.normal(0., 0.2, x.shape)

>    >>> spectrum = Spectrum1D(flux=y*u.Jy, spectral_axis=x*u.um)

Reviewer: Maybe break this up into two code blocks, and have the first one be a matplotlib ``plot`` directive showing the spectrum so the user can actually see the spectrum that yields this table? (You can just copy-and-paste the idioms I used in getting started or analysis if you like)

Author: as I understand, this is now done as part of the previous comment

   >>> # Define a noise region for adding the uncertainty
   >>> noise_region = SpectralRegion(0*u.um, 3*u.um)

   >>> # Derivative method
   >>> lines = find_lines_derivative(spectrum, flux_threshold=0.75)

   >>> emission_lines = lines[lines['line_type'] == 'emission'] #doctest:+SKIP
         line_center    line_type line_center_index
              um                                   
      ----------------- --------- -----------------
      4.522613065326634  emission                90
      5.477386934673367  emission               109

   >>> absorption_lines = lines[lines['line_type'] == 'absorption'] #doctest:+SKIP
         line_center    line_type  line_center_index
              um                                    
      ----------------- ---------- -----------------
      8.190954773869347 absorption               163

> 

Reviewer: ```suggestion
While it might be surprising that these tables do not contain more information
about the lines, this is because the "toolbox" philosophy of `specutils` aims to
keep such functionality in separate distinct functions.  See :doc:`analysis` for
functions that can be used to fill out common line measurements more
completely.

```

Author: will add

Parameter Estimation
--------------------

Given a spectrum with a set of lines, the `~specutils.fitting.estimate_line_parameters`
can be called to estimate the `~astropy.modeling.Model` parameters given a spectrum.

For the `~astropy.modeling.functional_models.Gaussian1D`, 
`~astropy.modeling.functional_models.Voigt1D`, and
`~astropy.modeling.functional_models.Lorentz1D` models, there are predefined estimators for each 

> of the parameters. For all other models one must define the estimators (see example below).

Reviewer: ```suggestion
of the parameters. For all other models one must define the estimators (see example below).
Note that in many (most?) cases where another model is needed, it may be better to create 
your own template models tailored to your specific spectra and skip this function entirely.
```

Author: added

For example:

.. code-block:: python

   >>> import numpy as np
   >>> from astropy.modeling import models
   >>> import astropy.units as u
   >>> from specutils import Spectrum1D
   >>> from specutils.fitting import estimate_line_parameters

>    >>> # Define the spectrum

Reviewer: Why not use the same spectrum as above?  I think it might be clearer to the user if this section all uses one example spectrum.

Author: done. line finding and parameter estimation now all uses the same one.  The line fitting section below does not and I think that is fine as there are different types of fitting shown below.

   >>> np.random.seed(0)
   >>> x = np.linspace(0., 10., 200)
   >>> y = 3 * np.exp(-0.5 * (x - 6.3)**2 / 0.8**2)
   >>> y += np.random.normal(0., 0.2, x.shape)
   >>> spectrum = Spectrum1D(flux=y*u.Jy, spectral_axis=x*u.um)

   >>> # Estimate parameter Gaussian1D
   >>> g_init = estimate_line_parameters(spectrum, models.Gaussian1D()) #doctest:+SKIP
      Model: Gaussian1D
      Inputs: ('x',)
      Outputs: ('y',)
      Model set size: 1
      Parameters:
              amplitude            mean             stddev     
                  Jy                um                um       
          ----------------- ----------------- -----------------
          3.354169257846847 6.218588636687762 1.608040201005025

> An example where the parameter estimators are defined:

Reviewer: Maybe add a few more words here?  I.e., something about how you have to attach the estimators as `_constraints`, and that this is a behavior that may change in future versions of astropy and/or specutils (since it's a pretty ugly hack on some level).

Author: added

functions that can be used to fill out common line measurements more
completely.

Parameter Estimation
--------------------

Given a spectrum with a set of lines, the `~specutils.fitting.estimate_line_parameters`
can be called to estimate the `~astropy.modeling.Model` parameters given a spectrum.

For the `~astropy.modeling.functional_models.Gaussian1D`,
`~astropy.modeling.functional_models.Voigt1D`, and
`~astropy.modeling.functional_models.Lorentz1D` models, there are predefined estimators for each
of the parameters. For all other models one must define the estimators (see example below).
Note that in many (most?) cases where another model is needed, it may be better to create
your own template models tailored to your specific spectra and skip this function entirely.

For example, based on the spectrum defined above we can first select a region:

.. code-block:: python

   >>> from specutils import SpectralRegion
   >>> from specutils.fitting import estimate_line_parameters
   >>> from specutils.manipulation import extract_region

   >>> sub_region = SpectralRegion(4*u.um, 5*u.um)
   >>> sub_spectrum = extract_region(spectrum, sub_region)

> Then estimate the line  parameters it it for a Gaussian line profile::

Reviewer 2: @eteq just found this typo, but I don't know how to fix it (it it)

yielding a single composite model result (which can be split back into its
components if desired).

Line Finding
------------

There are two techniques implemented in order to find emission and/or absorption
lines in a `~specutils.Spectrum1D` spectrum.

The first technique is `~specutils.fitting.find_lines_threshold` that will
find lines by thresholding the flux based on a factor applied to the
spectrum uncertainty.  The second technique is
`~specutils.fitting.find_lines_derivative` that will find the lines based
on calculating the derivative and then thresholding based on it.  Both techniques 
return an `~astropy.table.QTable` that contains columns ``line_center``, 
``line_type`` and ``line_center_index``.

Given we start with a spectrum:

> .. code-block:: python

Reviewer: I don't think the code-block is necessary given the ``:include-source:`` below?

              um                                    
      ----------------- ---------- -----------------
      8.190954773869347 absorption               163

While it might be surprising that these tables do not contain more information
about the lines, this is because the "toolbox" philosophy of ``specutils`` aims to
keep such functionality in separate distinct functions.  See :doc:`analysis` for
functions that can be used to fill out common line measurements more
completely.

Parameter Estimation
--------------------

Given a spectrum with a set of lines, the `~specutils.fitting.estimate_line_parameters`
can be called to estimate the `~astropy.modeling.Model` parameters given a spectrum.

For the `~astropy.modeling.functional_models.Gaussian1D`, 
`~astropy.modeling.functional_models.Voigt1D`, and
`~astropy.modeling.functional_models.Lorentz1D` models, there are predefined estimators for each 
of the parameters. For all other models one must define the estimators (see example below).
Note that in many (most?) cases where another model is needed, it may be better to create
your own template models tailored to your specific spectra and skip this function entirely.

> For example, based on the spectrum defined above:

Reviewer: ```suggestion
For example, based on the spectrum defined above we can first select a region
in the spectrum corresponding to a line:
```

.. code-block:: python

   >>> from specutils import SpectralRegion
   >>> from specutils.fitting import estimate_line_parameters
   >>> from specutils.manipulation import extract_region

   >>> sub_region = SpectralRegion(4*u.um, 5*u.um)
   >>> sub_spectrum = extract_region(spectrum, sub_region)

>    >>> # Estimate parameter Gaussian1D

Reviewer: ```suggestion
Then estimate the line  parameters it it for a Gaussian line profile::
```

>    >>> g_init = estimate_line_parameters(sub_spectrum, models.Gaussian1D()) #doctest:+SKIP

Reviewer: Why skip the doctest here?  Definitely we want to check this so the user sees the right thing!
```suggestion
   >>> g_init = estimate_line_parameters(sub_spectrum, models.Gaussian1D())
   >>> g_init
```

      Model: Gaussian1D
      Inputs: ('x',)
      Outputs: ('y',)
      Model set size: 1
      Parameters:
              amplitude            mean             stddev
                  Jy                um                um
          ------------------ ---------------- ------------------
          1.1845669151078486 4.57517271067525 0.3015075376884422

If an `~astropy.modeling.Model` is used that does not have the predefined parameter
estimators, or if one wants to use different parameter estimators then one can create
a dictionary where the key is the parameter name and the value is a lambda
function that operates on a spectrum. For example if one wants to estimate the line

> parmaeters of a `~astropy.modeling.functional_models.MexicanHat1D` one can

Reviewer: (typo)
```suggestion
parameters of a `~astropy.modeling.functional_models.MexicanHat1D` one can
```

define the ``estimators`` dictionary and attach in the model's ``_constraints`` 

> dictionary.  (This behavior may change in future versions of astropy or specutils).

Reviewer: ```suggestion
dictionary.  (Although be aware the use of ``_constraints`` may change in
future versions of astropy or specutils, as this is a work around for a known
limitation in `astropy.modeling`).
```
